### PR TITLE
XPath: Do not accept whitespace characters other than #x20, #x9, #xD, and #xA & fix normalize-space()

### PR DESCRIPTION
**Pull request by Ahmad-S792**
#### 7e1bc9e89bee15c3828b75aa012d3d398de57ae3
<pre>
XPath: Do not accept whitespace characters other than #x20, #x9, #xD, and #xA &amp; fix normalize-space()

<a href="https://bugs.webkit.org/show_bug.cgi?id=255841">https://bugs.webkit.org/show_bug.cgi?id=255841</a>

Reviewed by Chris Dumez.

This patch aligns WebKit with Blink / Chromium and Gecko / Firefox.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/01fd8522eaf35abdfa468f79333b350df96e1b62">https://chromium.googlesource.com/chromium/src.git/+/01fd8522eaf35abdfa468f79333b350df96e1b62</a> &amp;
<a href="https://chromium.googlesource.com/chromium/src.git/+/66c3b1b0ce745473f6478e1fa042d71f00ee9d47">https://chromium.googlesource.com/chromium/src.git/+/66c3b1b0ce745473f6478e1fa042d71f00ee9d47</a>

This patch aligns WebKit with web-spec [1] &amp; [2] and also fixes XML white spaces handling; #x20, #x9, #xD, and #xA.
We incorrectly handled other white spaces such as U+3000.

[1] <a href="https://www.w3.org/TR/1999/REC-xpath-19991116/#function-normalize-space">https://www.w3.org/TR/1999/REC-xpath-19991116/#function-normalize-space</a>
[2] <a href="https://www.w3.org/TR/1999/REC-xpath-19991116/#NT-ExprWhitespace">https://www.w3.org/TR/1999/REC-xpath-19991116/#NT-ExprWhitespace</a>

* Source/WebCore/xml/XPathFunctions.cpp:
(FunNormalizeSpace::evaluate):
* Source/WebCore/xml/XPathUtil.cpp:
(isXMLSpace): Introduce new bool function
* Source/WebCore/xml/XPathUtil.h: new bool function definition
* Source/WebCore/xml/XPathParser.cpp:
(Parser::skipWS): Use new &apos;isXMLSpace&apos; function
* LayoutTests/imported/w3c/web-platform-tests/domxpath/lexical-structure-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-normalize-space-expected.txt: Rebaselined

Canonical link: <a href="https://commits.webkit.org/263311@main">https://commits.webkit.org/263311@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/6946f67af8470406ece668a3f8ba3a548a843134

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4190 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4307 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4425 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5657 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4435 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4181 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4413 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4272 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4657 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4251 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4426 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3778 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5650 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1921 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3755 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5942 "7 flakes 249 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3748 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3821 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5338 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4227 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3428 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3739 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3753 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1034 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/7842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4037 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/xml/XPathUtil.cpp
Comment (annevk) [inline Source/WebCore/xml/XPathUtil.cpp]: Use `' '`, `'\t'` and such as we do elsewhere in the code?

(This also matches the definition of "HTTP space", might provide opportunity for further consolidation down the line.)

Comment (Ahmad-S792) [reply Source/WebCore/xml/XPathUtil.cpp]: Sure, I will fix it once EWS finishes and also will fix style error. If needed any help, I will ask. :-)

Comment (Ahmad-S792) [reply Source/WebCore/xml/XPathUtil.cpp]: Done.

File: Source/WebCore/xml/XPathFunctions.cpp
Comment (aproskuryakov) [inline Source/WebCore/xml/XPathFunctions.cpp]: This seems way less readable than the original.

File: Source/WebCore/xml/XPathUtil.cpp
Comment (aproskuryakov) [inline Source/WebCore/xml/XPathUtil.cpp:77]: XPath code seems like the wrong place fro this generic function. I know that Anne and Darin have been discussing related cleanup recently, not sure how closely that's related.

File: Source/WebCore/xml/XPathUtil.cpp
Comment (aproskuryakov) [inline Source/WebCore/xml/XPathUtil.cpp]: This is equivalent to `return character == ' ' || character == '\t' || character == '\r' || character == '\n'`.

File: Source/WebCore/xml/XPathUtil.h
Comment (aproskuryakov) [inline Source/WebCore/xml/XPathUtil.h]: Why is this include needed?

File: Source/WebCore/xml/XPathUtil.cpp
Comment (cdumez) [reply Source/WebCore/xml/XPathUtil.cpp]: This is a frequent performance optimization we have in our whitespace functions though. I'd keep as-is

File: Source/WebCore/xml/XPathFunctions.cpp
Comment (cdumez) [reply Source/WebCore/xml/XPathFunctions.cpp]: I agree, I'd rather we keep the explicit `if` branch.

File: Source/WebCore/xml/XPathUtil.cpp
Comment (cdumez) [reply Source/WebCore/xml/XPathUtil.cpp:77]: @annevk do you have a suggestion for a better location for this function?

File: Source/WebCore/xml/XPathUtil.cpp
Comment (aproskuryakov) [reply Source/WebCore/xml/XPathUtil.cpp]: Surprised if the compiler doesn't optimize it away. Have we ever confirmed that it changes generated code?

Comment (cdumez) [reply Source/WebCore/xml/XPathUtil.cpp]: Just verified out of curiosity and llvm does generate the same code with -O2 (tried both arm64 and x86_64):
```
isXMLSpace1(char):                       // @isXMLSpace1(char)
        mov     x8, #9728                       // =0x2600
        and     w9, w0, #0xff
        movk    x8, #1, lsl #32
        cmp     w9, #33
        cset    w9, lo
        lsr     x8, x8, x0
        and     w0, w9, w8
        ret
isXMLSpace2(char):                       // @isXMLSpace2(char)
        mov     x8, #9728                       // =0x2600
        and     w9, w0, #0xff
        movk    x8, #1, lsl #32
        cmp     w9, #33
        cset    w9, lo
        lsr     x8, x8, x0
        and     w0, w9, w8
        ret
```

@Ahmad-S792 Let's simplify the code then.

File: Source/WebCore/xml/XPathUtil.h
Comment (Ahmad-S792) [reply Source/WebCore/xml/XPathUtil.h]: I added it to use Unicode characters similar to other files but removing it does compile as well. So I will remove it.

File: Source/WebCore/xml/XPathUtil.cpp
Comment (Ahmad-S792) [reply Source/WebCore/xml/XPathUtil.cpp]: Will do in next push.

File: Source/WebCore/xml/XPathUtil.cpp
Comment (annevk) [reply Source/WebCore/xml/XPathUtil.cpp:77]: `ASCIICType.h` is a good location I think, following the recent move of `isHTMLSpace` (plus renaming) there, but then the name should be either `isHTTPWhitespace` or `isASCIIWhitespaceMinusFormFeed`/`isASCIIWhitespaceMinusFF`.  We'll primarily use it for HTTP, MIME types, and such, but apparently also for some XML-related features that are not covered by libxml.

Comment (Ahmad-S792) [reply Source/WebCore/xml/XPathUtil.cpp:77]: Will wait for @annevk's response and will work accordingly.

File: Source/WebCore/xml/XPathUtil.h
Comment (annevk) [reply Source/WebCore/xml/XPathUtil.h]: You only need this if you actually need to use one of the constants defined in that file.

File: Source/WebCore/xml/XPathFunctions.cpp
Comment (Ahmad-S792) [reply Source/WebCore/xml/XPathFunctions.cpp]: Done.

File: Source/WebCore/xml/XPathUtil.cpp
Comment (Ahmad-S792) [reply Source/WebCore/xml/XPathUtil.cpp]: Done.

File: Source/WebCore/xml/XPathUtil.h
Comment (Ahmad-S792) [reply Source/WebCore/xml/XPathUtil.h]: Done.

File: Source/WebCore/xml/XPathUtil.cpp
Comment (Ahmad-S792) [reply Source/WebCore/xml/XPathUtil.cpp:77]: I think `isASCIIWhitespaceMinusFormFeed` is clearer name to me.

Comment (Ahmad-S792) [reply Source/WebCore/xml/XPathUtil.cpp:77]: @aproskuryakov & @cdumez - I am able to do this in local build but my PR Git setup is bad and if I rebased then it will mess-up. I am happy to do follow-up to fix address this. My local branch is outdated and don't have @annevk's PR, so it will lead to merge conflict, if I do and rebasing tends to also mess-up this commit. Or I can close this PR and do new with all changes in one go.

Comment (Ahmad-S792) [reply Source/WebCore/xml/XPathUtil.cpp:77]: Only issue is that @annevk has few open PRs, so I can do once all of them are merged and remove 'XMLSpace' from these functions and add them to 'ASCIICType.h' as suggested later.

**Review by cdumez — APPROVED** [review]
LGTM

**Pull request merged**
